rfcs/20240228-graph-api-ocl-runtime/README.md

```diff
@@ -0,0 +1,272 @@
+    return stream(c_stream);
+}
+
+}
+}
+```
+
+#### Execution
+
+Corresponding to the `sycl::event` returned by compiled partition execution
+under SYCL runtime, for OpenCL runtime, we need to expose new APIs to return
+`cl_event`.
+
+Below C APIs will be defined and exposed when the library is built with OpenCL
+GPU runtime.
+
+```c
+/// Execute a compiled partition with ocl runtime.
+///
+/// @param compiled_partition The handle of target compiled_partition.
+/// @param stream The stream used for execution
+/// @param num_inputs The number of input tensors
+/// @param inputs A list of input tensors
+/// @param num_outputs The number of output tensors
+/// @param outputs A non-empty list of output tensors
+/// @param deps Optional handle of list with `cl_event` dependencies.
+/// @param ndeps Number of dependencies.
+/// @param ocl_event The handle of cl_event returned by the API.
+/// @returns #dnnl_success on success and a status describing the
+///     error otherwise.
+dnnl_status_t DNNL_API dnnl_graph_ocl_interop_compiled_partition_execute(
+        const_dnnl_graph_compiled_partition_t compiled_partition,
+        dnnl_stream_t stream, size_t num_inputs,
+        const_dnnl_graph_tensor_t *inputs, size_t num_outputs,
+        const_dnnl_graph_tensor_t *outputs, const cl_event *deps, int ndeps, cl_event *ocl_event);
+```
+
+C++ API in namespace `dnnl::graph::ocl_interop::`.
+
+```c++
+namespace dnnl {
+namespace graph {
+namespace ocl_interop {
+
+/// Executes a compiled partition in a specified stream and returns a OCL
+/// event.
+///
+/// @param c_partition Compiled partition to execute.
+/// @param astream Stream object to run over
+/// @param inputs Arguments map.
+/// @param outputs Arguments map.
+/// @param deps Optional vector with `cl_event` dependencies.
+/// @returns Output event.
+inline cl_event execute(compiled_partition &c_partition, stream &astream,
+        const std::vector<tensor> &inputs, std::vector<tensor> &outputs,
+        const std::vector<cl_event> &deps = {}) {
+    // call the C API dnnl_graph_ocl_interop_compiled_partition_execute to
+    // implement the logic.
+}
+
+}
+}
+}
+```
+
+### Testing
+
+Benchdnn graph driver needs to be enhanced to support OpenCL GPU runtime
+validation [[#3]][3] by respecting the GPU runtime build flag and call OpenCL
+interop API or SYCL interop API respectively.
+
+Since oneDNN has the same fusion capability for SYCL runtime and OpenCL runtime
+on GPU, the existing input files [[#4]][4] designed for SYCL GPU validation should be
+enough and can be reused to validate OpenCL GPU.
+
+### Limitations
+
+- Following the convention of SYCL GPU runtime in Graph API, only USM is
+  supported for OpenCL GPU runtime.
+- Primitive selection (`DNNL_ENABLE_PRIMITIVE`) and workload selection
+  (`DNNL_ENABLE_WORKLOAD`) are not supported by Graph API as before. We mention
+  them here as the features were initially requested by OpenVINO for primitive
+  API.
```

Comment:
What about the binary size impact of enabling graph api? This configs are added because we are constantly pushed for smaller binary size.

Reply:
For dynamic library, enabling Graph API (`-DONEDNN_BUILD_GRAPH=ON`) increases ~4MB to the binary size. It's almost constant over different CPU/GPU runtime combinations as the implementation code is general.

For example, a typical default build (CPU_RUNTIME=OMP/GPU_RUNTIME=NONE) with GCC 11.4 on my Ubuntu machine. The binary size of `libdnnl.so` is increased from 59710128 Bytes to 64126976 Bytes with Graph API enabled.